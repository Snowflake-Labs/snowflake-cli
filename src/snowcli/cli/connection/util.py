import logging
import typer
from click.exceptions import ClickException
from snowflake.connector import SnowflakeConnection
from snowflake.connector.cursor import DictCursor

LOCAL_DEPLOYMENT: str = "us-west-2"

log = logging.getLogger(__name__)


class MissingConnectionHostError(ClickException):
    def __init__(self, conn: SnowflakeConnection):
        super().__init__(
            f"The connection host ({conn.host}) was missing or not in "
            "the expected format "
            "(<account>.<deployment>.snowflakecomputing.com)"
        )


def get_deployment(conn: SnowflakeConnection) -> str:
    """
    Determines the deployment this connection refers to; useful
    to generate URLs that point to Snowsight.
    """
    if not conn.host:
        raise MissingConnectionHostError(conn)

    host_parts = conn.host.split(".")
    if host_parts[-1] == "local":
        return LOCAL_DEPLOYMENT

    if len(host_parts) != 6:
        raise MissingConnectionHostError(conn)

    return ".".join(host_parts[1:4])


def get_account(conn: SnowflakeConnection) -> str:
    """
    Determines the account that this connection refers to.
    """
    if conn.account:
        return conn.account

    if not conn.host:
        raise MissingConnectionHostError(conn)

    host_parts = conn.host.split(".")
    return host_parts[0]


def make_snowsight_url(conn: SnowflakeConnection, path: str) -> str:
    """Returns a URL on the correct Snowsight instance for the connected account."""
    try:
        *_, cursor = conn.execute_string(
            f"select system$get_snowsight_host()", cursor_class=DictCursor
        )
        snowsight_host = cursor.fetchone()["SYSTEM$GET_SNOWSIGHT_HOST()"]
    except Exception as e:
        # if we cannot determine the host, assume we're on prod
        snowsight_host = "https://app.snowflake.com"

    deployment = get_deployment(conn)
    account = get_account(conn)
    path_with_slash = path if path.startswith("/") else f"/{path}"
    return f"{snowsight_host}/{deployment}/{account}{path_with_slash}"
