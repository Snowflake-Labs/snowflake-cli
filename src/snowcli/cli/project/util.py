import re
import os
from typing import List


def clean_identifier(input):
    """
    Removes characters that cannot be used in an unquoted identifier,
    converting to lowercase as well.
    """
    return re.sub(r"[^a-z0-9_$]", "", f"{input}".lower())


def generate_user_env(username: str) -> dict:
    return {
        "USER": username,
    }


def first_set_env(*keys: List[str]):
    getenv = lambda i: os.getenv(keys[i], getenv(i + 1) if i + 1 < len(keys) else None)
    return getenv(0)


def get_env_username() -> str | None:
    return first_set_env("USER", "USERNAME", "LOGNAME")
