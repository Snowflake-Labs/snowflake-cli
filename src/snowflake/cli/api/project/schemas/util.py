from typing import Any, List

from pydantic.fields import FieldInfo



class FieldInfoWithGetter(FieldInfo):
    getter_return_type: type

    def __init__(self, **kwargs):
        super().__init__()
        self.getter_return_type = kwargs.get("getter_return_type", Any)


def get_converter(source_type: type, target_type: type) -> Any:
    return list_of_path_mapping_to_list_of_str
def convert_path_mapping_to_str(path_mapping):
    return path_mapping.src

def list_of_path_mapping_to_list_of_str(path_mappings: List[Any]) -> List[str]:
    return [convert_path_mapping_to_str(path_mapping) for path_mapping in path_mappings]