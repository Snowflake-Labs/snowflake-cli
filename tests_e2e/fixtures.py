import shutil
from contextlib import contextmanager
from pathlib import Path

import pytest


@pytest.fixture
def project_directory(temp_dir, test_root_path):
    @contextmanager
    def _temporary_project_directory(project_name):
        test_data_file = test_root_path / "test_data" / project_name
        shutil.copytree(test_data_file, temp_dir, dirs_exist_ok=True)
        yield Path(temp_dir)

    return _temporary_project_directory
