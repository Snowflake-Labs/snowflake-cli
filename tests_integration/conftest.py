from __future__ import annotations

import functools
import pytest
import toml

from pathlib import Path
from snowcli.cli import app
from tempfile import NamedTemporaryFile
from typer import Typer
from typer.testing import CliRunner

TEST_DIR = Path(__file__).parent


@pytest.fixture(scope="session")
def test_snowcli_config():
    test_config = TEST_DIR / "config/test.toml"
    config = toml.load(test_config)
    config["snowsql_config_path"] = str(TEST_DIR / "config/connection_configs.toml")
    with NamedTemporaryFile(suffix=".toml", mode="w+") as fh:
        toml.dump(config, fh)
        fh.flush()
        yield fh.name


class SnowCLIRunner(CliRunner):
    def __init__(self, app: Typer, test_snowcli_config: str):
        super().__init__()
        self.app = app
        self.test_snowcli_config = test_snowcli_config

    @functools.wraps(CliRunner.invoke)
    def invoke(self, *a, **kw):
        return super().invoke(self.app, *a, **kw)

    def invoke_with_config(self, *args, **kwargs):
        return self.invoke(
            ["--config-file", self.test_snowcli_config, *args[0]], **kwargs
        )


@pytest.fixture
def runner(test_snowcli_config):
    return SnowCLIRunner(app, test_snowcli_config)
