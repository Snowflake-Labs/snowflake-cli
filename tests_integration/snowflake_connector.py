from __future__ import annotations

import os
import pytest
import uuid

from snowflake import connector
from snowcli.exception import EnvironmentVariableNotFoundError

_ENV_PARAMETER_PREFIX = "SNOWFLAKE_CONNECTIONS_INTEGRATION"


@pytest.fixture(scope="session")
def create_database():
    database_name = f"db_{uuid.uuid4().hex}"

    config = {
        "application": "INTEGRATION_TEST",
        "account": _get_from_env("ACCOUNT", True),
        "user": _get_from_env("USER", True),
        "password": _get_from_env("PASSWORD", True),
        "schema": _get_from_env("SCHEMA", False, "PUBLIC"),
        "host": _get_from_env("HOST", False),
        "role": _get_from_env("ROLE", False),
    }
    config = {k: v for k, v in config.items() if v is not None}
    connection = connector.connect(**config)

    connection.execute_string(f"create database {database_name}")
    os.environ[f"{_ENV_PARAMETER_PREFIX}_DATABASE"] = database_name

    yield database_name

    connection.execute_string(f"drop database {database_name}")
    del os.environ[f"{_ENV_PARAMETER_PREFIX}_DATABASE"]


@pytest.fixture(scope="session")
def snowflake_session(create_database):
    config = {
        "application": "INTEGRATION_TEST",
        "account": _get_from_env("ACCOUNT", True),
        "user": _get_from_env("USER", True),
        "password": _get_from_env("PASSWORD", True),
        "host": _get_from_env("HOST", False),
        "database": _get_from_env("DATABASE", False),
        "role": _get_from_env("ROLE", False),
        "schema": "PUBLIC",
        "warehouse": _get_from_env("WAREHOUSE", False),
    }
    config = {k: v for k, v in config.items() if v is not None}
    connection = connector.connect(**config)
    yield connection
    connection.close()


def _get_from_env(parameter_name: str, required: bool, default=None) -> str | None:
    env_value = os.environ.get(f"{_ENV_PARAMETER_PREFIX}_{parameter_name}")
    if required and not env_value:
        if default is None:
            raise EnvironmentVariableNotFoundError(
                f"{_ENV_PARAMETER_PREFIX}_{parameter_name}"
            ) from None
        return default
    return env_value
