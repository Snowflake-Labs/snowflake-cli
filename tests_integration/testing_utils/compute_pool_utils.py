import math
import time

import pytest
from snowflake.connector import SnowflakeConnection

from tests_integration.conftest import SnowCLIRunner
from tests_integration.test_utils import contains_row_with, not_contains_row_with
from tests_integration.testing_utils.assertions.test_result_assertions import (
    assert_that_result_is_successful_and_executed_successfully,
)


class ComputePoolTestSetup:
    def __init__(
        self,
        runner: SnowCLIRunner,
        snowflake_session: SnowflakeConnection,
    ):
        self.runner = runner
        self.snowflake_session = snowflake_session


class ComputePoolTestSteps:
    def __init__(self, setup: ComputePoolTestSetup):
        self._setup = setup

    def create_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "spcs",
                "compute-pool",
                "create",
                compute_pool_name,
                "--min-nodes",
                1,
                "--family",
                "CPU_X64_XS",
            ]
        )
        assert result.json, result.output
        assert "status" in result.json
        assert (
            f"Compute Pool {compute_pool_name.upper()} successfully created."
            in result.json["status"]  # type: ignore
        )

    def list_should_return_compute_pool(self, compute_pool_name) -> None:
        result = self._execute_list()
        assert contains_row_with(result.json, {"name": compute_pool_name.upper()})

    def list_should_not_return_compute_pool(self, compute_pool_name: str) -> None:
        result = self._execute_list()
        assert not_contains_row_with(result.json, {"name": compute_pool_name.upper()})

    def describe_should_return_compute_pool(self, compute_pool_name: str) -> None:
        result = self._execute_describe(compute_pool_name)
        assert result.json
        assert len(result.json) == 1
        assert result.json[0]["name"] == compute_pool_name.upper()

    def drop_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "drop",
                "compute-pool",
                compute_pool_name,
            ],
        )
        assert result.json
        assert len(result.json) == 1
        assert result.json[0] == {  # type: ignore
            "status": f"{compute_pool_name.upper()} successfully dropped."
        }

    def stop_all_on_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            ["spcs", "compute-pool", "stop-all", compute_pool_name]
        )
        assert_that_result_is_successful_and_executed_successfully(result, is_json=True)

    def suspend_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            ["spcs", "compute-pool", "suspend", compute_pool_name]
        )
        assert_that_result_is_successful_and_executed_successfully(result, is_json=True)

    def resume_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            ["spcs", "compute-pool", "resume", compute_pool_name]
        )
        assert_that_result_is_successful_and_executed_successfully(result, is_json=True)

    def wait_until_compute_pool_is_idle(self, compute_pool_name: str) -> None:
        self._wait_until_compute_pool_reaches_state(compute_pool_name, "IDLE", 300)

    def wait_until_compute_pool_is_suspended(self, compute_pool_name: str) -> None:
        self._wait_until_compute_pool_reaches_state(compute_pool_name, "SUSPENDED", 60)

    def _wait_until_compute_pool_reaches_state(
        self, compute_pool_name: str, target_state: str, max_duration: int
    ):
        assert max_duration > 0
        max_counter = math.ceil(max_duration / 10)
        target_state = target_state.upper()
        for i in range(max_counter):
            status = self._execute_describe(compute_pool_name)
            if contains_row_with(status.json, {"state": target_state}):
                return
            time.sleep(10)
        status = self._execute_describe(compute_pool_name)

        error_message = f"Compute pool {compute_pool_name} didn't reach target state '{target_state}' in {max_duration} seconds. Current state is '{status.json['state']}'"
        pytest.fail(error_message)

    def _execute_describe(self, compute_pool_name: str):
        return self._setup.runner.invoke_with_connection_json(
            ["object", "describe", "compute-pool", compute_pool_name]
        )

    def _execute_list(self):
        return self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "list",
                "compute-pool",
            ],
        )
