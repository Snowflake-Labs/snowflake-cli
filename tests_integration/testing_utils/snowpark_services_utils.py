import json
import math
import time
from textwrap import dedent
from typing import Union

import pytest
from snowflake.connector import SnowflakeConnection

from tests_integration.conftest import SnowCLIRunner
from tests_integration.test_utils import contains_row_with, not_contains_row_with
from tests_integration.testing_utils.assertions.test_result_assertions import (
    assert_that_result_is_successful_and_output_json_contains,
    assert_that_result_is_successful_and_output_json_equals,
)


class SnowparkServicesTestSetup:
    def __init__(
        self,
        runner: SnowCLIRunner,
        snowflake_session: SnowflakeConnection,
        test_root_path,
    ):
        self.runner = runner
        self.snowflake_session = snowflake_session
        self.test_root_path = test_root_path


class SnowparkServicesTestSteps:
    compute_pool = "snowcli_compute_pool"
    database = "snowcli_db"
    schema = "public"
    container_name = "echo-test"

    def __init__(self, setup: SnowparkServicesTestSetup):
        self._setup = setup

    def create_service(self, service_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "spcs",
                "service",
                "create",
                service_name,
                "--compute-pool",
                self.compute_pool,
                "--spec-path",
                self._get_spec_path(),
                "--database",
                self.database,
                "--schema",
                self.schema,
            ],
        )
        assert_that_result_is_successful_and_output_json_equals(
            result, {"status": f"Service {service_name.upper()} successfully created."}
        )

    def status_should_return_service(self, service_name: str) -> None:
        result = self._execute_status(service_name)
        assert_that_result_is_successful_and_output_json_contains(
            result,
            {"containerName": self.container_name, "serviceName": service_name.upper()},
        )

    def logs_should_return_service_logs(self, service_name: str) -> None:
        result = self._setup.runner.invoke_with_connection(
            [
                "spcs",
                "service",
                "logs",
                service_name,
                "--container-name",
                self.container_name,
                "--instance-id",
                "0",
                "--database",
                self.database,
                "--schema",
                self.schema,
            ],
        )
        assert result.output
        # Assert this instead of full payload due to log coloring
        assert service_name in result.output
        assert '"GET /healthcheck HTTP/1.1" 200 -' in result.output

    def list_should_return_service(self, service_name: str) -> None:
        result = self._execute_list()
        assert contains_row_with(result.json, {"name": service_name.upper()})

    def list_should_not_return_service(self, service_name: str) -> None:
        result = self._execute_list()
        assert not_contains_row_with(result.json, {"name": service_name.upper()})

    def describe_should_return_service(self, service_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "describe",
                "service",
                f"{self.database}.{self.schema}.{service_name}",
            ],
        )
        assert result.json
        assert result.json[0]["name"] == service_name.upper()  # type: ignore

    def drop_service(self, service_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "drop",
                "service",
                f"{self.database}.{self.schema}.{service_name}",
            ],
        )
        assert result.json[0] == {  # type: ignore
            "status": f"{service_name.upper()} successfully dropped."
        }

    def wait_until_service_is_ready(self, service_name: str) -> None:
        self._wait_until_service_reaches_state(service_name, "READY", 900)

    def _wait_until_service_reaches_state(
        self, service_name: str, target_status: Union[str, dict], max_duration: int
    ):
        assert max_duration > 0
        max_counter = math.ceil(max_duration / 10)
        if isinstance(target_status, str):
            target_status = {"status": target_status}
        for i in range(max_counter):
            status = self._execute_status(service_name)
            if contains_row_with(status.json, target_status):
                return
            elif contains_row_with(status.json, {"status": "FAILED"}):
                describe = self._setup.runner.invoke_with_connection_json(
                    ["object", "describe", "service", service_name]
                )
                pytest.fail(
                    dedent(
                        f"""
                    {service_name} service failed before reaching target state:
                    {json.dumps(target_status)}
                    current state:
                    {json.dumps(status)}
                    current describe:
                    {json.dumps(describe)}
                    """
                    )
                )
            time.sleep(10)
        status = self._execute_status(service_name)

        error_message = f"""
{service_name} service didn't reach target state in {max_duration} seconds.
target:
{json.dumps(target_status)}
current:
{json.dumps(status.json)}
"""
        pytest.fail(error_message)

    def _execute_status(self, service_name: str):
        return self._setup.runner.invoke_with_connection_json(
            [
                "spcs",
                "service",
                "status",
                service_name,
                "--database",
                self.database,
                "--schema",
                self.schema,
            ],
        )

    def _execute_list(self):
        return self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "list",
                "service",
            ],
        )

    def _get_spec_path(self):
        return f"{self._setup.test_root_path}/spcs/spec/spec.yml"
