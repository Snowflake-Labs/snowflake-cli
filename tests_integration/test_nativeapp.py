import pytest
import os
import uuid
from pathlib import Path
from contextlib import contextmanager

from snowcli.cli.project.util import generate_user_env
from tests.project.fixtures import *
from tests.testing_utils.fixtures import temp_dir
from tests_integration.snowflake_connector import test_database, snowflake_session
from tests_integration.test_utils import (
    row_from_snowflake_session,
    contains_row_with,
)


USER_NAME = f"user_{uuid.uuid4().hex}"
TEST_ENV = generate_user_env(USER_NAME)


@contextmanager
def pushd(dir: Path):
    cwd = os.getcwd()
    os.chdir(dir)
    try:
        yield dir
    finally:
        os.chdir(cwd)


@pytest.mark.integration
def test_nativeapp_init_run_without_modifications(
    runner,
    snowflake_session,
    temp_dir,
):
    project_name = "myapp"
    result = runner.invoke_with_config(
        ["app", "init", project_name],
        env=TEST_ENV,
    )
    assert result.exit_code == 0

    with pushd(Path(os.getcwd(), project_name)):
        result = runner.invoke_integration(
            ["app", "run"],
            env=TEST_ENV,
        )
        assert result.exit_code == 0

        try:
            # app + package exist
            package_name = f"{project_name}_pkg_{USER_NAME}".upper()
            app_name = f"{project_name}_{USER_NAME}".upper()
            assert contains_row_with(
                row_from_snowflake_session(
                    snowflake_session.execute_string(
                        f"show application packages like '{package_name}'",
                    )
                ),
                dict(name=package_name),
            )
            assert contains_row_with(
                row_from_snowflake_session(
                    snowflake_session.execute_string(
                        f"show applications like '{app_name}'",
                    )
                ),
                dict(name=app_name),
            )

        finally:
            # make sure we always delete the app
            result = runner.invoke_integration(
                ["app", "teardown"],
                env=TEST_ENV,
            )
            assert result.exit_code == 0


@pytest.mark.integration
@pytest.mark.parametrize("project_definition_files", ["integration"], indirect=True)
def test_nativeapp_run_existing(
    runner,
    snowflake_session,
    project_definition_files: List[Path],
):
    project_name = "integration"
    dir = project_definition_files[0].parent
    with pushd(dir):
        result = runner.invoke_integration(
            ["app", "run"],
            env=TEST_ENV,
        )
        print(result.output.encode("utf-8") if result.output else None)
        assert result.exit_code == 0

        try:
            # app + package exist
            package_name = f"{project_name}_pkg_{USER_NAME}".upper()
            app_name = f"{project_name}_{USER_NAME}".upper()
            assert contains_row_with(
                row_from_snowflake_session(
                    snowflake_session.execute_string(
                        f"show application packages like '{package_name}'",
                    )
                ),
                dict(name=package_name),
            )
            assert contains_row_with(
                row_from_snowflake_session(
                    snowflake_session.execute_string(
                        f"show applications like '{app_name}'"
                    )
                ),
                dict(name=app_name),
            )

            # sanity checks
            assert contains_row_with(
                row_from_snowflake_session(
                    snowflake_session.execute_string(
                        f"select count(*) from {app_name}.core.shared_view"
                    )
                ),
                {"COUNT(*)": 1},
            )
            test_string = "TEST STRING"
            assert contains_row_with(
                row_from_snowflake_session(
                    snowflake_session.execute_string(
                        f"call {app_name}.core.echo('{test_string}')"
                    )
                ),
                {"ECHO": test_string},
            )
        finally:
            # make sure we always delete the app
            result = runner.invoke_integration(
                ["app", "teardown"],
                env=TEST_ENV,
            )
            assert result.exit_code == 0
