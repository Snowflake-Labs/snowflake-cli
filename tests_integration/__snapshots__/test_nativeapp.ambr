# serializer version: 1
# name: test_nativeapp_deploy_directory_no_recursive
  '''
  ╭─ Error ──────────────────────────────────────────────────────────────────────╮
  │ /private/var/folders/d9/qwvk5h3s1cx9vrqpyjhsxy6m0000gn/T/tmp0dm1i_tl/app     │
  │ root/app/nested is a directory. Add the -r flag to deploy directories.       │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  '''
# ---
